### ISSUE TITLE
[RLlib] Bug when using RNNs, time_major and PyTorch

### ISSUE THREAD

**Issue by raphaelavalos** [ray-project/ray#13075]
Hello,
When using TrajectoryView the padding of the sequences doesn't take into account `time_major`. The padded sequences are then given to `add_time_dimension` which returns the wrong output if `time_major=True`.

What currently happens: [A, B, B, C] -> [A, P, B, B, C, P] -> [[A,P, B], [B,C,P]]
While it should be: [[A, B,C], [A,B,P]]

I have managed to fix it by reshaping it as if it were `time_major=False` and transposing the first two axes. There might be a cheaper way to do it by modifying the padding function.

https://github.com/ray-project/ray/blob/85f1716a1f011501c9d5f8bf2043c0c3f356ba95/rllib/policy/rnn_sequencing.py#L175-L180

@sven1977 If you could take a look :)

**Comment by sven1977** [comment]
Thanks for filing this issue @raphaelavalos ! Could you provide a minimal reproduction script so I can debug?
Or point me to the example/test that fails iff you set "_time_major=True".

**Comment by raphaelavalos** [comment]
Hello @sven1977,

I need to get up to speed with the newer version of ray as many things changed since the beginning of December (congrats on all the amazing work btw) before I can give you a reproduction script because my codes don't run on the newer version. So I will try to explain the bug again:
The padding happens in the same way whether `time_major` is true or not and it is done by adding elements at the end of the sequences so that every sequence is of the same size. The resulting tensor can be seen as a concatenation of the padded sequences (i.e. the first T elements correspond the first padded sequence). Then `add_time_dimension`:
* if `time_major=False`, reshapes the tensor to `(B, T, *)` which is correct because the first T elements are the first batch.
* if `time_major=True`, reshapes the tensor to `(T, B, *)` which is incorrect because the first B elements are not the first time-steps of every sequence.

In my solution I run `add_time_dimension` with `time_major=False` then I do a permutation of the first two axes.

I hope it is more clear,
Best

**Comment by XuehaiPan** [comment]
In PyTorch, RNN cells are in time-major by default, for example:

```python
lstm = nn.LSTM(num_outputs, cell_size)  # time-major
lstm = nn.LSTM(num_outputs, cell_size, batch_first=not time_major)  # time-major unless specified
```

the default value for argument `batch_first` is `False` in PyTorch.

------

In RLlib, batches for RNNs are stored in batch-major format by default. Users should use `config["_time_major"] = True` to have time-major batches.

```python
batch.size()  # -> (B, T, *)  // batch-major
batch.size()  # -> (T, B, *)  // time-major
```

For RNNs in RLlib, users implement `model.forward_rnn` method instead of `model.forward`. In `rnn_model.forward`, function `add_time_dimension` wraps inputs in shape `(B x T, *)` to `(B, T, *)` or `(T, B, *)`.

https://github.com/ray-project/ray/blob/c4636c7c05262a67991e67f277e5052c8d9083e4/rllib/models/torch/recurrent_net.py#L67-L87

But in `add_time_dimension`:

https://github.com/ray-project/ray/blob/c4636c7c05262a67991e67f277e5052c8d9083e4/rllib/policy/rnn_sequencing.py#L189-L198

The `reshape` operation produces the correct result when in batch-major since the data are stored in row-major format. But the wrong results when in time-major.

- In batch-major, given a batch index `i`, `batch[i, 0:T, ...]` gives sequential data.

    ```python
    import numpy as np
    import torch
    from ray.rllib.policy.rnn_sequencing import add_time_dimension
    
    B = 8
    T = 10
    F = 1
    
    input = torch.from_numpy(np.arange(B * T * F).reshape(B * T, -1))
    input.size()  # -> torch.Size([80, 1]) // (B x T, *)
    
    batch_major = add_time_dimension(input, max_seq_len=T, framework="torch", time_major=False)
    batch_major.size()  # -> torch.Size([8, 10, 1]) // (B, T, *)
    
    # Given a simple:
    #           B  T  *
    batch_major[0, :, ...]
    # -> tensor([[0], [1], [2], [3], [4], [5], [6], [7], [8], [9]])
    ```

- In time-major, the results are non-sequential along the time dimension. Given a batch index `i`, `batch[0:T, i, ...]` is not sequential data.

    ```python
    time_major = add_time_dimension(input, max_seq_len=T, framework="torch", time_major=True)
    time_major.size()  # -> torch.Size([10, 8, 1]) // (T, B, *)
    
    # Given a simple:
    #           T  B  *
    time_major[:, 0, ...]
    # -> tensor([[0], [8], [16], [24], [32], [40], [48], [56], [64], [72]])
    ```

The correct result should use `transpose` or `permute`:

```python
# In add_time_dimemsion:
time_major = input.reshape((T, B, *))
# i.e.
batch_major = input.reshape((B, T, *))
time_major = batch_major.reshape((T, B, *))  # non-sequential along the time dimention

# Correct approach:
batch_major = input.reshape((B, T, *))
time_major = batch_major.transpose(0, 1)
```
